File: plugins/ossl/engine/lib/src/common/hash.rs
// Copyright (C) Microsoft Corporation. All rights reserved.

use std::ffi::c_uint;

use mcr_api_resilient::DigestKind;
use openssl_rust::safeapi::error::OpenSSLError;
use openssl_rust::safeapi::error::OpenSSLResult;
use openssl_rust::NID_sha1 as NID_SHA1;
use openssl_rust::NID_sha256 as NID_SHA256;
use openssl_rust::NID_sha384 as NID_SHA384;
use openssl_rust::NID_sha512 as NID_SHA512;

/// Get the hash type for sign/verify from NID
pub(crate) fn azihsm_hash_type(nid: c_uint) -> OpenSSLResult<Option<DigestKind>> {
    match nid {
        0 => Ok(None),
        NID_SHA1 => Ok(Some(DigestKind::Sha1)),
        NID_SHA256 => Ok(Some(DigestKind::Sha256)),
        NID_SHA384 => Ok(Some(DigestKind::Sha384)),
        NID_SHA512 => Ok(Some(DigestKind::Sha512)),
        _ => Err(OpenSSLError::HashNotSupported),
    }
}

/// Get the NID from the hash type
pub(crate) fn openssl_hash_nid(digest_kind: Option<DigestKind>) -> c_uint {
    match digest_kind {
        None => 0,
        Some(DigestKind::Sha1) => NID_SHA1,
        Some(DigestKind::Sha256) => NID_SHA256,
        Some(DigestKind::Sha384) => NID_SHA384,
        Some(DigestKind::Sha512) => NID_SHA512,
    }
}

File: plugins/ossl_engine/tests/lib/Hash.cpp
// Copyright (C) Microsoft Corporation. All rights reserved.

#include "AziHsmEngine.hpp"
#include "AziHsmHash.hpp"
#include <stdexcept>
#include <vector>
#include <openssl/rand.h>
#include <openssl/evp.h>
#include <openssl/sha.h>

AziHsmEvpMdCtx::AziHsmEvpMdCtx(EVP_PKEY_CTX *pctx)
    : pctx(pctx)
{
    this->ctx = EVP_MD_CTX_create();
    if (this->ctx == nullptr)
    {
        throw std::runtime_error("Could not create EVP_MD_CTX");
    }

    if (pctx != nullptr)
    {
        EVP_MD_CTX_set_pkey_ctx(this->ctx, pctx);
    }
}

AziHsmEvpMdCtx::~AziHsmEvpMdCtx()
{
    if (this->pctx != nullptr)
    {
        EVP_PKEY_CTX_free(this->pctx);
    }
    EVP_MD_CTX_destroy(this->ctx);
}

EVP_MD_CTX *AziHsmEvpMdCtx::getCtx()
{
    return this->ctx;
}

AziHsmShaHash::AziHsmShaHash(AziHsmShaHashType type)
{
    switch (type)
    {
    case AziHsmShaHashType::SHA512:
    {
        this->type = (EVP_MD *)EVP_sha512();
        break;
    }
    case AziHsmShaHashType::SHA384:
    {
        this->type = (EVP_MD *)EVP_sha384();
        break;
    }
    case AziHsmShaHashType::SHA256:
    {
        this->type = (EVP_MD *)EVP_sha256();
        break;
    }
    case AziHsmShaHashType::SHA1:
    {
        this->type = (EVP_MD *)EVP_sha1();
        break;
    }
    default:
    {
        throw std::runtime_error("Could not detect SHA type");
    }
    }
}

int AziHsmShaHash::getSize()
{
    return EVP_MD_size(this->type);
}

int AziHsmShaHash::getNid()
{
    return EVP_MD_type(this->type);
}

EVP_MD *AziHsmShaHash::getType()
{
    return this->type;
}

EVP_MD_CTX *AziHsmShaHash::getCtx()
{
    return this->ctx.getCtx();
}

AziHsmShaHashType AziHsmShaHash::getHashType()
{
    int nid = this->getNid();
    switch (nid)
    {
    case NID_sha1:
        return AziHsmShaHashType::SHA1;
    case NID_sha256:
        return AziHsmShaHashType::SHA256;
    case NID_sha384:
        return AziHsmShaHashType::SHA384;
    case NID_sha512:
        return AziHsmShaHashType::SHA512;
    default:
        throw std::runtime_error("Invalid NID passed");
    }
}

std::vector<unsigned char> AziHsmShaHash::hashData(std::vector<unsigned char> &data)
{
    if (!EVP_DigestInit_ex(this->getCtx(), this->type, NULL))
    {
        throw std::runtime_error("Could not initialize EVP digest");
    }

    if (!EVP_DigestUpdate(this->getCtx(), data.data(), data.size()))
    {
        throw std::runtime_error("Could not update digest");
    }

    size_t len = EVP_MD_CTX_size(this->getCtx());
    std::vector<unsigned char> hash(len);

    if (!EVP_DigestFinal_ex(this->getCtx(), hash.data(), (unsigned int *)&len))
    {
        throw std::runtime_error("Could not finalize digest");
    }

    return hash;
}
